contrib/mesos/pkg/minion/server.go

```diff
@@ -0,0 +1,268 @@
+	log "github.com/golang/glog"
+	"github.com/kardianos/osext"
+	"github.com/spf13/pflag"
+	"gopkg.in/natefinch/lumberjack.v2"
+)
+
+type MinionServer struct {
+	KubeletExecutorServer *exservice.KubeletExecutorServer
+	privateMountNS        bool
+	hks                   hyperkube.Interface
+	clientConfig          *client.Config
+	kmBinary              string
+	done                  chan struct{} // closed when shutting down
+	exit                  chan error    // to signal fatal errors
+
+	logMaxSize      resource.Quantity
+	logMaxBackups   int
+	logMaxAgeInDays int
+
+	runProxy     bool
+	proxyLogV    int
+	proxyBindall bool
+}
+
+func NewMinionServer() *MinionServer {
+	s := &MinionServer{
+		KubeletExecutorServer: exservice.NewKubeletExecutorServer(),
+
+		privateMountNS: false, // disabled until Docker supports customization of the parent mount namespace
+		done:           make(chan struct{}),
+		exit:           make(chan error),
+
+		logMaxSize:      config.DefaultLogMaxSize(),
+		logMaxBackups:   config.DefaultLogMaxBackups,
+		logMaxAgeInDays: config.DefaultLogMaxAgeInDays,
+
+		runProxy: true,
+	}
+
+	// cache this for later use
+	binary, err := osext.Executable()
+	if err != nil {
+		log.Fatalf("failed to determine currently running executable: %v", err)
+	}
+	s.kmBinary = binary
+
+	return s
+}
+
+// filterArgsByFlagSet returns a list of args which are parsed by the given flag set
+// and another list with those which do not match
+func filterArgsByFlagSet(args []string, flags *pflag.FlagSet) ([]string, []string) {
+	matched := []string{}
+	notMatched := []string{}
+	for _, arg := range args {
+		err := flags.Parse([]string{arg})
+		if err != nil {
+			notMatched = append(notMatched, arg)
+		} else {
+			matched = append(matched, arg)
+		}
+	}
+	return matched, notMatched
+}
+
+func (ms *MinionServer) launchProxyServer() {
+	bindAddress := "0.0.0.0"
+	if !ms.proxyBindall {
+		bindAddress = ms.KubeletExecutorServer.Address.String()
+	}
+	args := []string{
+		fmt.Sprintf("--bind-address=%s", bindAddress),
+		fmt.Sprintf("--v=%d", ms.proxyLogV),
+		"--logtostderr=true",
+	}
+
+	if ms.clientConfig.Host != "" {
+		args = append(args, fmt.Sprintf("--master=%s", ms.clientConfig.Host))
+	}
+
+	ms.launchHyperkubeServer(hyperkube.KM_PROXY, &args, "proxy.log")
+}
+
+func (ms *MinionServer) launchExecutorServer() {
+	allArgs := os.Args[1:]
+
+	// filter out minion flags, leaving those for the executor
+	executorFlags := pflag.NewFlagSet("executor flags", pflag.ContinueOnError)
+	executorFlags.SetOutput(ioutil.Discard)
+	ms.AddExecutorFlags(executorFlags)
+	executorArgs, _ := filterArgsByFlagSet(allArgs, executorFlags)
+
+	// run executor and quit minion server when this exits cleanly
+	err := ms.launchHyperkubeServer(hyperkube.KM_EXECUTOR, &executorArgs, "executor.log")
+	if err != nil {
+		// just return, executor will be restarted on error
+		log.Error(err)
+		return
+	}
+
+	log.Info("Executor exited cleanly, stopping the minion")
+	ms.exit <- nil
```

Comment:
added a comment. Closing ms.exit leads to panics for the go routines writing to it. We had that before and changed it.